rust_snuba/benches/processors.rs

```diff
@@ -0,0 +1,364 @@
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use divan::counter::ItemsCount;
+use once_cell::sync::Lazy;
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use rust_arroyo::backends::local::broker::LocalBroker;
+use rust_arroyo::backends::local::LocalConsumer;
+use rust_arroyo::backends::storages::memory::MemoryMessageStorage;
+use rust_arroyo::backends::{Consumer, ConsumerError};
+use rust_arroyo::processing::strategies::run_task_in_threads::ConcurrencyConfig;
+use rust_arroyo::processing::strategies::ProcessingStrategyFactory;
+use rust_arroyo::processing::{RunError, StreamProcessor};
+use rust_arroyo::types::{Partition, Topic};
+use rust_arroyo::utils::clock::SystemClock;
+use rust_snuba::{
+    ClickhouseConfig, ConsumerStrategyFactory, MessageProcessorConfig, StorageConfig,
+};
+use uuid::Uuid;
+
+fn main() {
+    divan::main();
+}
+
+const MSG_COUNT: usize = 10_000;
+
+#[divan::bench]
+fn functions(bencher: divan::Bencher) {
+    run_bench(
+        bencher,
+        functions_payload,
+        "FunctionsMessageProcessor",
+        "TODO",
+    );
+}
+
+#[divan::bench]
+fn profiles(bencher: divan::Bencher) {
+    run_bench(
+        bencher,
+        profiles_payload,
+        "ProfilesMessageProcessor",
+        "TODO",
+    );
+}
+
+#[divan::bench]
+fn querylog(bencher: divan::Bencher) {
+    run_bench(bencher, querylog_payload, "QuerylogProcessor", "TODO");
+}
+
+#[divan::bench]
+fn spans(bencher: divan::Bencher) {
+    run_bench(bencher, spans_payload, "SpansMessageProcessor", "TODO");
+}
+
+fn run_bench(
+    bencher: divan::Bencher,
+    make_payload: fn() -> KafkaPayload,
+    processor: &str,
+    schema: &str,
+) {
+    bencher
+        .counter(ItemsCount::new(MSG_COUNT))
+        .with_inputs(|| {
+            (
+                create_consumer(make_payload, MSG_COUNT),
+                create_factory(processor, schema),
+            )
+        })
+        .bench_local_values(|((topic, consumer), factory)| {
+            let mut processor = StreamProcessor::new(consumer, factory);
+            processor.subscribe(topic);
+
+            loop {
+                let res = processor.run_once();
+                if matches!(res, Err(RunError::Poll(ConsumerError::EndOfPartition))) {
+                    // FIXME: this pretty much means that we *polled* the partition to the end,
+                    // it does not mean that we actually *committed* everything
+                    // (aka we finished actually processing everything).
+                    return;
+                }
+            }
+            // FIXME: this seems to deadlock?
+            //processor.shutdown();
```

Comment:
I guess you are wanting to run this benchmark with the StreamProcessor here, since it's more realistic. Though it won't fully work until the assignment callback locking problem is fixed. If you wanted to benchmark the strategies in isolation and submit messages directly to them, that could work already.